tianshou/policy/modelfree/redq.py

```diff
@@ -0,0 +1,207 @@
+from copy import deepcopy
+from typing import Any, Dict, Optional, Tuple, Union
+
+import numpy as np
+import torch
+from torch.distributions import Independent, Normal
+
+from tianshou.data import Batch, ReplayBuffer, to_torch_as
+from tianshou.exploration import BaseNoise
+from tianshou.policy import DDPGPolicy
+
+
+class REDQPolicy(DDPGPolicy):
```

Comment:
Why don't you inherit `SACPolicy` to match the setup in the paper as well as your MBPO implementation?

Reply:
Inheriting `SACPolicy` requires changes of currently working code `sac.py` as the 'critics1, ...' parameters are required now. I'm not sure if this is acceptable at this moment.

The same thing happens on `EnsembleMLP` in `tianshou.utils.net.common.py`. There will be less duplicated code if I can pass the `ensemble_size` as a parameter of `EnsembleLinear` to
https://github.com/thu-ml/tianshou/blob/bc53ead273f6f9d3788a78ecc739249eeb96b8c6/tianshou/utils/net/common.py#L15-L19

Reply:
I see. Since `SACPolicy` is a sub-class of `DDPGPolicy`, you only go up one level. I think that's a good compromise. Thanks for the explanation!

Reply:
Regarding `EmsembleMLP` and friends, how about something like this:

```
def miniblock(
    input_size: int,
    output_size: int = 0,
    norm_layer: Optional[ModuleType] = None,
    activation: Optional[ModuleType] = None,
    linear_layer: Type[nn.Linear] = nn.Linear,
    ensemble_size: int = 1,
) -> List[nn.Module]:
    """Construct a miniblock with given input/output-size, norm layer and \
    activation."""
    layers: List[nn.Module] = [linear_layer(input_size, output_size) if ensemble_size == 1 else linear_layer(ensemble_size, input_size, output_size)]
```

Reply:
I was considering
```
def miniblock(
    input_size: int,
    output_size: int = 0,
    norm_layer: Optional[ModuleType] = None,
    activation: Optional[ModuleType] = None,
    linear_layer: Type[nn.Linear] = nn.Linear,
    **layer_args
) -> List[nn.Module]:
    """Construct a miniblock with given input/output-size, norm layer and \
    activation."""
    layers: List[nn.Module] = [linear_layer(input_size, output_size, **layer_args)]
```
and then change the parameter order of `EnsembleLinear`, so that it can be compatible with some other fancy layers in the future.

Reply:
Wait. There might be simpler solutions. Keep `EnsembleLinear`, then define a lambda function to pass in to `MLP` or `miniblock()` as `linear_layer`:

```
def linear_layer(input_size, output_size):
  return EnsembleLinear(ensemble_size, input_size, output_size)
```

Similar to https://github.com/thu-ml/tianshou/blob/c25926dd8f5b6179f7f76486ee228982f48b4469/examples/atari/atari_network.py#L110-L114